src/Microsoft.AspNet.OData.Shared/Extensions/ODataPathExtensions.cs

```diff
@@ -41,6 +41,59 @@ internal static class ODataPathExtensions
             return keys;
         }
 
+        public static ODataPathSegment GetLastNonKeySegment(this ODataPath path)
+        {
+            if (path == null)
+            {
+                throw Error.ArgumentNull(nameof(path));
+            }
+
+            if (path.Count == 1)
+            {
+                return path.LastSegment;
+            }
+
+            if (path.LastSegment is TypeSegment)
+            {
+                ODataPath modifiedPath = path.TrimEndingTypeSegment();
+
+                return modifiedPath.LastSegment;
+            }
+
+            return path.LastSegment;
+        }
+
+        public static ODataPathSegment GetLastNonTypeNonKeySegment(this ODataPath path)
+        {
+            if (path == null)
+            {
+                throw Error.ArgumentNull(nameof(path));
+            }
+
+            if (path.Count == 1)
+            {
+                return path.LastSegment;
+            }
+
+            ODataPath modifiedPath = path;
+
+            while (modifiedPath.LastSegment is TypeSegment || modifiedPath.LastSegment is KeySegment)
+            {
+                if (modifiedPath.LastSegment is TypeSegment)
+                {
+                    modifiedPath = modifiedPath.TrimEndingTypeSegment();
```

Comment:
having a short description of what happens here will be very helpful. You can have an example of `LastSegment` and the result after the `TrimEndingTypeSegment` is called.